File: cpp/dynamic_programming/knapsack_01.cpp
// 0/1 Knapsack Problem
// Time Complexity: O(n*W), Space Complexity: O(n*W)

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    // Method 1: Memoization
    int solveMemo(vector<int>& wt, vector<int>& val, int n, int W, vector<vector<int>>& dp) {
        if (n == 0 || W == 0) return 0;
        if (dp[n][W] != -1) return dp[n][W];
        
        if (wt[n-1] <= W) {
            return dp[n][W] = max(
                val[n-1] + solveMemo(wt, val, n-1, W-wt[n-1], dp),
                solveMemo(wt, val, n-1, W, dp)
            );
        } else {
            return dp[n][W] = solveMemo(wt, val, n-1, W, dp);
        }
    }
    
    // Method 2: Tabulation
    int knapsack(vector<int>& wt, vector<int>& val, int n, int W) {
        vector<vector<int>> dp(n+1, vector<int>(W+1, 0));
        
        for (int i = 1; i <= n; i++) {
            for (int w = 1; w <= W; w++) {
                if (wt[i-1] <= w) {
                    dp[i][w] = max(val[i-1] + dp[i-1][w-wt[i-1]], dp[i-1][w]);
                } else {
                    dp[i][w] = dp[i-1][w];
                }
            }
        }
        return dp[n][W];
    }
    
    // Method 3: Space Optimized
    int knapsackOptimized(vector<int>& wt, vector<int>& val, int n, int W) {
        vector<int> prev(W+1, 0), curr(W+1, 0);
        
        for (int i = 1; i <= n; i++) {
            for (int w = 1; w <= W; w++) {
                if (wt[i-1] <= w) {
                    curr[w] = max(val[i-1] + prev[w-wt[i-1]], prev[w]);
                } else {
                    curr[w] = prev[w];
                }
            }
            prev = curr;
        }
        return prev[W];
    }
};

int main() {
    Solution sol;
    int n, W;
    cout << "Enter number of items: ";
    cin >> n;
    
    vector<int> wt(n), val(n);
    cout << "Enter weights: ";
    for (int i = 0; i < n; i++) cin >> wt[i];
    
    cout << "Enter values: ";
    for (int i = 0; i < n; i++) cin >> val[i];
    
    cout << "Enter knapsack capacity: ";
    cin >> W;
    
    cout << "Maximum value: " << sol.knapsack(wt, val, n, W) << endl;
    return 0;
}

File: python/dynamic_programming/knapsack_01.py
# 0/1 Knapsack Problem
# Time Complexity: O(n*W), Space Complexity: O(n*W)

from typing import List

class Solution:
    # Method 1: Memoization
    def solve_memo(self, wt: List[int], val: List[int], n: int, W: int, dp: dict) -> int:
        if n == 0 or W == 0:
            return 0
        if (n, W) in dp:
            return dp[(n, W)]
        
        if wt[n-1] <= W:
            dp[(n, W)] = max(
                val[n-1] + self.solve_memo(wt, val, n-1, W-wt[n-1], dp),
                self.solve_memo(wt, val, n-1, W, dp)
            )
        else:
            dp[(n, W)] = self.solve_memo(wt, val, n-1, W, dp)
        
        return dp[(n, W)]
    
    # Method 2: Tabulation
    def knapsack(self, wt: List[int], val: List[int], n: int, W: int) -> int:
        dp = [[0] * (W + 1) for _ in range(n + 1)]
        
        for i in range(1, n + 1):
            for w in range(1, W + 1):
                if wt[i-1] <= w:
                    dp[i][w] = max(val[i-1] + dp[i-1][w-wt[i-1]], dp[i-1][w])
                else:
                    dp[i][w] = dp[i-1][w]
        
        return dp[n][W]
    
    # Method 3: Space Optimized
    def knapsack_optimized(self, wt: List[int], val: List[int], n: int, W: int) -> int:
        prev = [0] * (W + 1)
        curr = [0] * (W + 1)
        
        for i in range(1, n + 1):
            for w in range(1, W + 1):
                if wt[i-1] <= w:
                    curr[w] = max(val[i-1] + prev[w-wt[i-1]], prev[w])
                else:
                    curr[w] = prev[w]
            prev = curr[:]
        
        return prev[W]


if __name__ == "__main__":
    sol = Solution()
    n = int(input("Enter number of items: "))
    wt = list(map(int, input("Enter weights: ").split()))
    val = list(map(int, input("Enter values: ").split()))
    W = int(input("Enter knapsack capacity: "))
    
    print(f"Maximum value: {sol.knapsack(wt, val, n, W)}")
